src/app/ReadClient.cpp

```diff
@@ -0,0 +1,326 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines read client for a CHIP Interaction Data model
+ *
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/ReadClient.h>
+
+namespace chip {
+namespace app {
+
+CHIP_ERROR ReadClient::Init(Messaging::ExchangeManager * apExchangeMgr, EventCallback const aEventCallback, void * const apAppState)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    // Error if already initialized.
+    VerifyOrExit(mpExchangeMgr == nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(mpExchangeCtx == nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    Reset();
+    mpAppState     = apAppState;
+    mpExchangeMgr  = apExchangeMgr;
+    mpExchangeCtx  = nullptr;
+    mEventCallback = aEventCallback;
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+void ReadClient::Reset()
+{
+    mpAppState = nullptr;
+    ClearExistingExchangeContext();
+    MoveToState(kState_Initialized);
+    return;
+}
+
+void ReadClient::Shutdown()
+{
+    VerifyOrExit(mState != kState_Uninitialized, );
+    mpAppState = nullptr;
+    ClearExistingExchangeContext();
+    mpExchangeMgr = nullptr;
+    MoveToState(kState_Uninitialized);
+
+exit:
+    return;
+}
+
+#if CHIP_DETAIL_LOGGING
```

Comment:
you can add this around the `switch` and have no else branch.

Reply:
Done, thanks